core/test/utils/unsort_matrix.hpp

```diff
@@ -0,0 +1,168 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_TEST_UTILS_UNSORT_MATRIX_HPP_
+#define GKO_CORE_TEST_UTILS_UNSORT_MATRIX_HPP_
+
+#include <algorithm>
+#include <random>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/base/iterator_factory.hpp"
+
+
+namespace gko {
+namespace test {
+
+
+// Plan for now: shuffle values and column indices to unsort the given matrix
+// without changing the represented matrix.
+template <typename ValueType, typename IndexType, typename RandomEngine>
+void unsort_matrix(matrix::Csr<ValueType, IndexType> *mtx,
+                   RandomEngine &&engine)
+{
+    using value_type = ValueType;
+    using index_type = IndexType;
+    auto size = mtx->get_size();
+    if (mtx->get_num_stored_elements() <= 0) {
+        return;
+    }
+    const auto &exec = mtx->get_executor();
+    auto master = mtx->get_executor()->get_master();
+
+    // If exec is not the master/host, extract the master and perform the
+    // unsorting there, followed by copying it back
+    if (exec != master) {
+        constexpr int max_depth{10};  // Max depth that is searched for master
+        bool actual_master_found{false};
+        for (int i = 0; i < max_depth; ++i) {
+            const auto new_master = master->get_master();
+            if (new_master == master) {
+                actual_master_found = true;
+                break;
+            }
+            master = new_master;
+        }
+        if (!actual_master_found) {
+            GKO_NOT_SUPPORTED(exec);
+        }
```

Comment:
Maybe I am missing something here, are there any executors with depth > 1 ? I think for all executors `exec->get_master()` should directly give you the master, right ?

Reply:
Usually, yes, but there is also nothing preventing us from doing something like:
`auto exec = gko::CudaExecutor::create(0, gko::CudaExecutor::create(0, gko::OmpExecutor::create()))`.
We don't guarantee that `get_master()` is actually a CPU executor.

Reply:
That should not happen, the `get_master` should indeed directly be the CPU side executor.

Reply:
I tried it inside an example, I can create such a chain as an executor, but the example fails during runtime (but it compiles without any issue).
Do you want me to remove this part since it should not be necessary? (The current implementation should work either way.)

Reply:
Yes, I think it is better to remove it. I dont think it makes sense to create an executor whose `get_master()` would be a non-CPU executor, as Terry mentions.